core/src/main/scala/org/bykn/bosatsu/PackageCustoms.scala

```diff
@@ -0,0 +1,171 @@
+    // OR the type is used
+    val impValues: SortedSet[(PackageName, Identifier)] =
+      (for {
+        imp <- pack.imports.iterator
+        impPack = imp.pack.name
+        if !impPack.isPredef // ignore predef
+        item <- imp.items.iterator
+        ref <- item.tag.iterator
+        name <- ref match {
+          case Constructor(_, _) | Referant.Value(_) =>
+            item.originalName :: Nil
+          case DefinedT(_) => Nil
+        }
+      } yield (impPack, name)).to(SortedSet)
+
+    val impTypes: SortedSet[(PackageName, Type.Const)] =
+      (for {
+        imp <- pack.imports.iterator
+        impPack = imp.pack.name
+        if !impPack.isPredef // ignore predef
+        item <- imp.items.iterator
+        ref <- item.tag.iterator
+        tpe <- ref match {
+          case Constructor(_, _) | Referant.Value(_) =>
+            Nil
+          case DefinedT(dt) =>
+            dt.toTypeConst :: Nil
+        }
+      } yield (impPack, tpe)).to(SortedSet)
+
+    if (impValues.isEmpty && impTypes.isEmpty) Chain.empty
+    else {
+      type VSet = Set[(PackageName, Identifier)]
+      type VState[X] = State[VSet, X]
+      val usedValuesSt: VState[Unit] =
+        pack.program.lets.traverse_ { case (_, _, te) =>
+          te.traverseUp {
+            case g@TypedExpr.Global(p, n, _, _) =>
+              State { s => (s + ((p, n)), g) }
+            case m @ Match(_, branches, _) => 
+              branches.traverse_ {
+                case (pat, _) =>
+                  pat.traverseStruct[VState, (PackageName, Identifier.Constructor)] { (n, parts) =>
+                    State.modify[VSet](_ + n) *> 
+                      parts.map { inner =>
+                        Pattern.PositionalStruct(n, inner)
+                      }
+                  }
+                  .void
+              }.as(m)
+            case te => Monad[VState].pure(te)
+          }
+        }
+
+      val usedValues = usedValuesSt.runS(Set.empty).value
+      
+
+      val usedTypes: Set[Type.Const] =
+        pack.program.lets.iterator.flatMap(
+          _._3.allTypes.flatMap(Type.constantsOf(_))
+        ).toSet
+
+      val unusedValues = impValues.filterNot { tup =>
+        tup match {
+          case (pn, cons @ Identifier.Constructor(_)) =>
+            // deal with the ambiguity of capital names
+            usedValues(tup) || usedTypes(Type.Const.Defined(pn, TypeName(cons)))
+          case _ =>
+            // no ambiguity for bindable
+            usedValues(tup)
+        }  
+      }
+      val unusedTypes = impTypes.filterNot { case (pn, t) =>
+        // deal with the ambiguity of capital names
+        usedTypes(t) || usedValues((pn, t.toDefined.name.ident))
+      }
+
+      val unusedValMap = unusedValues.groupByNes(_._1)
+      val unusedTypeMap = unusedTypes.groupByNes(_._1)
+      val unusedPacks = unusedValMap.keySet | unusedTypeMap.keySet
+
+      Chain.fromIterableOnce(
```

Comment:
all these errors are for a single package, we should go ahead and group them rather than forget that we have a set of errors of the same type from the same package.